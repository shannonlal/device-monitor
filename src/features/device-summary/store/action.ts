import { Action } from 'redux';
import { IDeviceDetailsModel } from '../../../interfaces/models';

// Actions Constants
export const GET_DEVICE_DETAILS_START = 'GET_DEVICE_DETAILS_START';
export const GET_DEVICE_DETAILS_SUCCESS = 'GET_DEVICE_DETAILS_START_SUCCESS';
export const GET_DEVICE_DETAILS_ERROR = 'GET_DEVICE_DETAILS_ERROR';

// Action Types
export interface IActionGetDeviceDetailsStart extends Action {
    type: 'GET_DEVICE_DETAILS_START';
}

export interface IActionGetDeviceDetailsSuccess extends Action {
    type: 'GET_DEVICE_DETAILS_START_SUCCESS';
    deviceDetails: IDeviceDetailsModel;
}

export interface IActionGetDeviceDetailsError extends Action {
    type: 'GET_DEVICE_DETAILS_ERROR';
    errorMessage: string;
}

export type DeviceDetailsActions =
    | IActionGetDeviceDetailsStart
    | IActionGetDeviceDetailsSuccess
    | IActionGetDeviceDetailsError;
// Action Functions
export function getDeviceDetailsStart(): IActionGetDeviceDetailsStart {
    return {
        type: GET_DEVICE_DETAILS_START,
    };
}

export function getDeviceDetailsSuccess(deviceDetails: IDeviceDetailsModel): IActionGetDeviceDetailsSuccess {
    return {
        type: GET_DEVICE_DETAILS_SUCCESS,
        deviceDetails,
    };
}

export function getDeviceDetailsError(errorMessage: string): IActionGetDeviceDetailsError {
    return {
        type: GET_DEVICE_DETAILS_ERROR,
        errorMessage,
    };
}
