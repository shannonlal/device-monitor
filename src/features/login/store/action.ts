import { Action } from 'redux';
import { IAuthorization, IUser } from '../../../interfaces/models';

// Actions Constants
export const AUTHORIZE_USER_START = 'AUTHORIZE_USER_START';
export const AUTHORIZE_USER_SUCCESS = 'AUTHORIZE_USER_SUCCESS';
export const AUTHORIZE_USER_ERROR = 'AUTHORIZE_USER_ERROR';

// Action Types
export interface IActionAuthorizeUserStart extends Action {
    type: 'AUTHORIZE_USER_START';
    auth: IAuthorization;
}

export interface IActionAuthorizeUserSuccess extends Action {
    type: 'AUTHORIZE_USER_SUCCESS';
    user: IUser;
}

export interface IActionAuthorizeUserError extends Action {
    type: 'AUTHORIZE_USER_ERROR';
    errorMessage: string;
}

export type AuthorizeUserActions = IActionAuthorizeUserStart | IActionAuthorizeUserSuccess | IActionAuthorizeUserError;

// Action Functions
export function authorizeUserStart(auth: IAuthorization): IActionAuthorizeUserStart {
    return {
        type: AUTHORIZE_USER_START,
        auth,
    };
}

export function authorizeUserSuccess(user: IUser): IActionAuthorizeUserSuccess {
    return {
        type: AUTHORIZE_USER_SUCCESS,
        user,
    };
}

export function getDeviceDetailsError(errorMessage: string): IActionAuthorizeUserError {
    return {
        type: AUTHORIZE_USER_ERROR,
        errorMessage,
    };
}
