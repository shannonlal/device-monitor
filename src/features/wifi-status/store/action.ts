import { Action } from 'redux';
import { IWifiSummary, IWifiDetails } from '../../../interfaces/models';

/************** Select Wifi  *****************/
// Actions Constants
export const SELECTED_WIFI_START = 'SELECTED_WIFI_START';
export const SELECTED_WIFI_SUCCESS = 'SELECTED_WIFI_SUCCESS';
export const SELECTED_WIFI_FAIL = 'SELECTED_WIFI_FAIL';

// Action Types
export interface IActionSelectedWifiStart extends Action {
    type: 'SELECTED_WIFI_START';
    ssid: string;
}

export interface IActionSelectedWifiSuccess extends Action {
    type: 'SELECTED_WIFI_SUCCESS';
    wifiDetails: IWifiDetails;
}

export interface IActionSelectedWifiFail extends Action {
    type: 'SELECTED_WIFI_FAIL';
    errorMessage: string;
}

export type SelectWifiActions = IActionSelectedWifiStart | IActionSelectedWifiSuccess | IActionSelectedWifiFail;

// Action Functions
export function selectWifiStart(ssid: string): IActionSelectedWifiStart {
    return {
        type: SELECTED_WIFI_START,
        ssid,
    };
}

export function selectWifiSuccess(wifiDetails: IWifiDetails): IActionSelectedWifiSuccess {
    return {
        type: SELECTED_WIFI_SUCCESS,
        wifiDetails,
    };
}

export function selectWifiFail(errorMessage: string): IActionSelectedWifiFail {
    return {
        type: SELECTED_WIFI_FAIL,
        errorMessage,
    };
}

/************** Get Avaialble  Wifi  *****************/
// Actions Constants
export const GET_AVAILABILE_WIFI_START = 'GET_AVAILABILE_WIFI_START';
export const GET_AVAILABILE_WIFI_SUCCESS = 'GET_AVAILABILE_WIFI_SUCCESS';
export const GET_AVAILABILE_WIFI_FAIL = 'GET_AVAILABILE_WIFI_FAIL';

// Action Types
export interface IActionGetAvailableWifiStart extends Action {
    type: 'GET_AVAILABILE_WIFI_START';
}

export interface IActionGetAvailableWifiSuccess extends Action {
    type: 'GET_AVAILABILE_WIFI_SUCCESS';
    wifiNetworks: IWifiSummary[];
}

export interface IActionGetAvailableWifiFail extends Action {
    type: 'GET_AVAILABILE_WIFI_FAIL';
    errorMessage: string;
}

export type GetAvailableWifiActions =
    | IActionGetAvailableWifiStart
    | IActionGetAvailableWifiSuccess
    | IActionGetAvailableWifiFail;

// Action Functions
export function getAvaialbleWifiStart(): IActionGetAvailableWifiStart {
    return {
        type: GET_AVAILABILE_WIFI_START,
    };
}

export function getAvaialbleWifiSuccess(wifiNetworks: IWifiSummary[]): IActionGetAvailableWifiSuccess {
    return {
        type: GET_AVAILABILE_WIFI_SUCCESS,
        wifiNetworks,
    };
}

export function getAvaialbleWifiFail(errorMessage: string): IActionGetAvailableWifiFail {
    return {
        type: GET_AVAILABILE_WIFI_FAIL,
        errorMessage,
    };
}
