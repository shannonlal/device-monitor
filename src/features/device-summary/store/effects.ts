import { Dispatch } from 'redux';
import { getDeviceDetailsStart, getDeviceDetailsSuccess, getDeviceDetailsError } from './action';
import { IDeviceDetailsModel } from '../../../interfaces/models';
import { getDeviceBridge } from '../../../drivers/DriverFactory';

/**
 * Will get the device data
 * @return
 */
const getDeviceData = async (): Promise<IDeviceDetailsModel> => {
    const deviceDetails: IDeviceDetailsModel = await getDeviceBridge().getDeviceInfo();

    return deviceDetails;
};

/**
 *
 */
export function getDeviceDetails() {
    return async (dispatch: Dispatch) => {
        dispatch(getDeviceDetailsStart());
        try {
            const deviceDetails: IDeviceDetailsModel = await getDeviceData();
            return dispatch(getDeviceDetailsSuccess(deviceDetails));
        } catch (err) {
            return dispatch(getDeviceDetailsError(err.message));
        }
    };
}
