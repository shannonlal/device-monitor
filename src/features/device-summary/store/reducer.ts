import { IDeviceDetailsModel } from '../../../interfaces/models';
import { DeviceDetailsActions } from './action';
export interface IDeviceDetailsState {
    state: string; // 'INIT', 'LOADING' | 'LOADED' | 'ERROR',
    deviceDetails?: IDeviceDetailsModel;
    errorMessage?: string;
}

export const initialState: IDeviceDetailsState = {
    state: 'INIT',
    errorMessage: '',
};

export function deviceDetailsReducer(
    state: IDeviceDetailsState = initialState,
    action: DeviceDetailsActions,
): IDeviceDetailsState {
    if (action.type === 'GET_DEVICE_DETAILS_START') {
        return {
            ...state,
            state: 'LOADING',
        };
    }
    if (action.type === 'GET_DEVICE_DETAILS_START_SUCCESS') {
        return {
            ...state,
            state: 'LOADED',
            deviceDetails: action.deviceDetails,
        };
    }
    if (action.type === 'GET_DEVICE_DETAILS_ERROR') {
        return {
            ...state,
            state: 'ERROR',
            errorMessage: action.errorMessage,
        };
    }
    return state;
}
