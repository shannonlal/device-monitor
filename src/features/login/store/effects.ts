import { Dispatch } from 'redux';
import { authorizeUserStart, authorizeUserSuccess, authorizeUserError  } from './action';
import { IAuthorization, IUser } from '../../../interfaces/models';
import LoginService from '../../../services/LoginService';
import { IResponse } from '../../../interfaces/response';

/**
 *
 */
export function authorizeUser(auth: IAuthorization) {
    return async (dispatch: Dispatch) => {
        dispatch(authorizeUserStart(auth));
        try {
            const loginService = new LoginService();
            const response: IResponse<IUser> = await loginService.login(auth);
            if (response.content) {
                const user: IUser = response.content;
                return dispatch(authorizeUserSuccess(user));
            } else {
                return dispatch(authorizeUserError(`Failed Authorization for user ${auth.eMail}`));
            }
        } catch (err) {
            return dispatch(authorizeUserError(err.message));
        }
    };
}
