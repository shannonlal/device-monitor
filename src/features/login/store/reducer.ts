import { IUser, IAuthorization } from '../../../interfaces/models';
import { AuthorizeUserActions } from './action';
export interface IAuthorizationState {
    state: string; // 'INIT', 'AUTHENTICATING' | 'AUTHENTICATED' | 'ERROR',
    auth?: IAuthorization;
    user?: IUser;
    errorMessage?: string;
}

export const initialState: IAuthorizationState = {
    state: 'INIT',
    errorMessage: '',
};

export function loginReducer(
    state: IAuthorizationState = initialState,
    action: AuthorizeUserActions,
): IAuthorizationState {
    if (action.type === 'AUTHORIZE_USER_START') {
        return {
            ...state,
            state: 'AUTHENTICATING',
            auth: action.auth,
        };
    }
    if (action.type === 'AUTHORIZE_USER_SUCCESS') {
        return {
            ...state,
            state: 'AUTHENTICATED',
            user: action.user,
        };
    }
    if (action.type === 'AUTHORIZE_USER_ERROR') {
        return {
            ...state,
            state: 'ERROR',
            errorMessage: action.errorMessage,
        };
    }
    return state;
}
