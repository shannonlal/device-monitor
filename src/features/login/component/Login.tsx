import React from 'react';
import { IonItem, IonInput, IonContent, IonLabel, IonButton } from '@ionic/react';
import { useForm, Controller } from 'react-hook-form';
import { IUser, IAuthorization } from '../../../interfaces/models';
import { useSelector, useDispatch } from 'react-redux';
import { authorizeUser } from '../store/effects';

const initialValues = {
    eMail: '',
    password: '',
};

interface IProps {
}

// Areas to focus on next
// 1. Add the new action, effects and reducer for login service
// 2. Integrate the service into the component
// 3. Implement Unit Tests for this.

export const Login: React.FC<IProps> = (): React.ReactElement => {
    const dispatch = useDispatch();

    useSelector((state) => {
        console.log('Here is the state', state);
    });
    const { control, handleSubmit, errors } = useForm<IAuthorization>({
        defaultValues: { ...initialValues },
        mode: 'onChange',
    });

    const onSubmit = (auth: IAuthorization) => {
        dispatch(authorizeUser(auth));
    };

    return (
        <IonContent>
            <form onSubmit={handleSubmit(onSubmit)}>
                <IonItem>
                    <IonLabel>Email</IonLabel>
                    <Controller
                        as={IonInput}
                        control={control}
                        onChangeName="onIonChange"
                        type="email"
                        onChange={([selected]) => {
                            console.log('eMail', selected.detail.value);
                            return selected.detail.value;
                        }}
                        name="eMail"
                        rules={{
                            required: true,
                            minLength: { value: 4, message: 'Must be 4 chars long' },
                        }}
                    />
                </IonItem>
                {errors.eMail && errors.eMail.type === 'required' && (
                    <div className="error">Your must enter your eMail.</div>
                )}
                <IonItem>
                    <IonLabel>Password</IonLabel>
                    <Controller
                        as={IonInput}
                        control={control}
                        onChangeName="onIonChange"
                        type="password"
                        onChange={([selected]) => {
                            console.log('fullName', selected.detail.value);
                            return selected.detail.value;
                        }}
                        name="password"
                        rules={{
                            required: true,
                            minLength: { value: 4, message: 'Must be 4 chars long' },
                        }}
                    />
                </IonItem>
                {errors.password && errors.password.type === 'required' && (
                    <div className="error">Your must enter your Password.</div>
                )}
                <IonButton type="submit" color="primary">
                    Primary
                </IonButton>
            </form>
        </IonContent>
    );
};

//export default DeviceDetails;
