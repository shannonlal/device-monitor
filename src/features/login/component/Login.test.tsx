import React from 'react';
//import { testRender, makeTestStore } from '../../../store/test-utils';
//import { loadSetsSuccess, setCurrentSet, zoomImage } from '../dataLayer';
import { useDispatch, useSelector } from 'react-redux';
import { Login } from './Login';
import { Action } from 'redux';
import { AuthorizeUserActions } from '../store/action';

import { render, fireEvent } from '@testing-library/react';

jest.mock('react-redux', () => ({
    ...jest.requireActual('react-redux'),
    useDispatch: jest.fn(),
    useSelector: jest.fn(),
}));

const useDispatchMock = useDispatch as jest.Mock;
const useSelectorMock = useSelector as jest.Mock;

const dispatchResultRecorder = {} as any;
const fakeDispatch = (action: AuthorizeUserActions) => {

    console.log( 'Calling Fake Dispatch', action.type);
    dispatchResultRecorder[action.type] = action;
};

useDispatchMock.mockImplementation(() => fakeDispatch);

describe('the Login component', () => {
    it('renders the login page', () => {
        const { asFragment } = render(<Login />);

        expect(asFragment()).toMatchSnapshot();
    });
});
