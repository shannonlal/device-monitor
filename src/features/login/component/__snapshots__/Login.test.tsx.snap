// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the Login component renders the login page 1`] = `
<DocumentFragment>
  <ion-content>
    <form>
      <ion-item>
        <ion-label>
          Email
        </ion-label>
        <ion-input
          name="eMail"
          type="email"
          value=""
        />
      </ion-item>
      <ion-item>
        <ion-label>
          Password
        </ion-label>
        <ion-input
          name="password"
          type="password"
          value=""
        />
      </ion-item>
      <ion-button
        color="primary"
        type="submit"
      >
        Primary
      </ion-button>
    </form>
  </ion-content>
</DocumentFragment>
`;
